provisioning/transport/amqp/src/SaslTpmHandler.cs

```diff
@@ -0,0 +1,183 @@
+            string sas = Convert.ToBase64String(sasBytes);
+
+            var responseBuffer = new byte [sas.Length * sizeof(char) + 1];
+            responseBuffer[0] = 0x0;
+            Buffer.BlockCopy(Encoding.UTF8.GetBytes(sas), 0, responseBuffer, 1, sas.Length);
+
+            var response = new SaslResponse { Response = new ArraySegment<byte>(responseBuffer) };
+            Negotiator.WriteFrame(response, true);
+
+        }
+
+        private void SaveEncodedNonceSegment(SaslChallenge saslChallenge)
+        {
+            var sequenceNumber = GetSequenceNumber(saslChallenge);
+            if (_nextSequenceNumber != sequenceNumber)
+                throw new AmqpException(AmqpErrorCode.InvalidField,
+                    $"Invalid sequence number, expected: {_nextSequenceNumber}, actual: {sequenceNumber}.");
+
+            string nonceSegment = Encoding.UTF8.GetString(saslChallenge.Challenge.Array, 1, saslChallenge.Challenge.Array.Length - 1);
+            _encodedNonceStringBuilder.Append(nonceSegment);
+            _nextSequenceNumber++;
+        }
+
+        private byte GetSequenceNumber(SaslChallenge saslChallenge)
+        {
+            return (byte) (saslChallenge.Challenge.Array[0] & SaslControlByteMask.SequenceNumber);
+        }
+
+        private void SendStorageRootKey()
+        {
+            var response = new SaslResponse {Response = new ArraySegment<byte>(CreateStorageRootKeyMessage())};
+            Negotiator.WriteFrame(response, true);
+        }
+
+        private byte[] CreateStorageRootKeyMessage()
+        {
+            var responseBuffer = new byte[_storageRootKey.Length + 1];
+            responseBuffer[0] = 0x0;
+            Buffer.BlockCopy(_storageRootKey, 0, responseBuffer, 1, _storageRootKey.Length);
+            return responseBuffer;
+        }
+
+        private SaslChallengeAction GetChallengeAction(SaslChallenge saslChallenge)
+        {
+            var challengeBytes = saslChallenge.Challenge.Array;
+
+            if (challengeBytes == null || challengeBytes.Length == 0)
+                throw new AmqpException(AmqpErrorCode.InvalidField, "Sasl challenge message is missing or empty.");
+
+            if (challengeBytes[0] == 0x0)
+                return SaslChallengeAction.First;
+
+            if ((challengeBytes[0] & SaslControlByteMask.InterimSegment) != 0x0)
+            {
+                if ((challengeBytes[0] & SaslControlByteMask.FinalSegment) != 0x0)
+                    return SaslChallengeAction.Final;
+
+                return SaslChallengeAction.Interim;
+            }
+
+            throw new AmqpException(AmqpErrorCode.InvalidField,
+                $"Sasl challenge control byte contains invalid data: {challengeBytes[0]:X}.");
+        }
+
+        public override void OnResponse(SaslResponse response)
+        {
+            //This method is only implemented on the server
+            var exception = new AmqpException(AmqpErrorCode.NotImplemented,
```

Comment:
Debug.Fail instead of throwing. Since it's not user actionable, I would hope the device doesn't crash because this got called for some reason by AMQP.